### PR TITLE
bpo-33462: Add __reversed__ to dict and dict views

### PR DESCRIPTION
**Pull request by remilapeyre**
[bpo-33462](https://www.bugs.python.org/issue33462) - Add `__reverse__` to `dict`

See https://bugs.python.org/issue33462

https://bugs.python.org/issue33462

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


When your account is ready, please add a comment in this pull request
and a Python core developer will remove the `CLA not signed` label
to make the bot check again.

Thanks again to your contribution and we look forward to looking at it!

**Comment by matrixise** [comment]
@remilapeyre Thank you so much for your PR but could you rename your PR with this format: "bpo-NNNN: TITLE".

Thank you

**Comment by matrixise** [comment]
and I forgot one thing, you need to have the CLA for a review. Thank you

File: Objects/dictobject.c
Comment (vstinner) [inline Objects/dictobject.c]: The PEP 7 now requires { ... } for if blocks.

File: Objects/dictobject.c
Comment (vstinner) [inline Objects/dictobject.c]: For new types, I now suggest to use the ".tp_dealloc = ...," syntax.

File: Lib/test/test_dict.py
Comment (selik) [inline Lib/test/test_dict.py]: Because integers have a special `hash()` behavior, perhaps a better test would be to use a few letters: "abcdefg". This shouldn't matter for the current dict implementation, but might make the test more robust against future implementation changes.

```
def test_reversed(self):
    keys = list('abcdefg')
    d = dict.fromkeys(keys)
    r = reversed(d)
    self.assertEqual(list(r), keys[::-1])
    self.assertRaises(StopIteration, next, r)
```

File: Lib/test/test_enumerate.py
Comment (selik) [inline Lib/test/test_enumerate.py]: Again, `dict.fromkeys('abcdefg')` might be better here than `{i: i for i in range(20)}`.

Historically, dicts using int keys ended up sorted by accident. It's hard to imagine this being a problem, but it might give a false positive if the dict implementation changes later.

**Comment by selik** [comment]
@remilapeyre There's a slight typo in the pull request title. It should be `__reversed__` not `__reverse__`.

**Comment by remilapeyre** [comment]
@selik  I updated the tests, thanks.

File: Objects/dictobject.c
Comment (methane) [inline Objects/dictobject.c:3349]: why?  When dict->ma_used is zero, no need to scan entries.

File: PC/python3.def
Comment (methane) [inline PC/python3.def]: Should these values be exposed by DLL?

File: Objects/dictobject.c
Comment (methane) [inline Objects/dictobject.c]: Can't we skip defining `.tp_getattro`?

File: Objects/dictobject.c
Comment (methane) [inline Objects/dictobject.c]: This seem wrong.

```
d = {'a':1, 'b':2, 'c':3}
del d['b']
```

ma_used is 2, and dk_nentries is 3.  dk_nentries-1 should be used.

File: Objects/dictobject.c
Comment (methane) [inline Objects/dictobject.c:3683]: Since we use C11 from Python 3.6, I prefer [DeclareVariablesAtFirstUse](http://wiki.c2.com/?DeclareVariablesAtFirstUse) style for new code.

File: Objects/dictobject.c
Comment (remilapeyre) [reply Objects/dictobject.c]: Hi @methane, thanks for the comments. I'm fixing the other issues. For this attribute, I'm not sure to understand precisely how `tp_getattro` is used and set it here because it was set on `PyDictKeys_Type`. As I understand it, it's the implementation of `getattr`. Is that right?

File: PC/python3.def
Comment (remilapeyre) [reply PC/python3.def]: Hi, if my memory is good (so I may absolutly be wrong) tests were not passing on windows before I added the types in this file. PyListRevIter_Type is here too so I thought those three new types may need to be here too but I'm not sure what puporse this file serves and if it is really necessary.

File: Objects/dictobject.c
Comment (methane) [reply Objects/dictobject.c]: When NULL is set to members, PyType_Ready fills them to default values.
And `PyObject_GenericGetAttr` is the default value.

When we skip specifying all defaults explicitly, programmers can focus on
meaningful part of the code.

Comment (remilapeyre) [reply Objects/dictobject.c]: Thank you so much for the clarification, I read `PyType_Ready` definition and it all makes sense now.

The documentation at https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_getattro for this field says 

>An optional pointer to the get-attribute function.
> 
> The signature is the same as for PyObject_GetAttr(). It is usually convenient to set this field to PyObject_GenericGetAttr(), which implements the normal way of looking for object attributes.

Would it makes sense to propose a change from *"It is usually convenient to set this field to PyObject_GenericGetAttr(), which implements the normal way of looking for object attributes."* to something like *"If not set, `PyType_Ready` will set this field to PyObject_GenericGetAttr(), which implements the normal way of looking for object attributes."* for this field and the ones alike?

File: Objects/dictnotes.txt
Comment (methane) [inline Objects/dictnotes.txt]: Oh, this note was wrong.  Please don't touch.  I will rewrite this paragraph.

File: Lib/_collections_abc.py
Comment (methane) [inline Lib/_collections_abc.py]: What is `isinstance({}, Reversible)` output without register?

Since Reversible uses `__subclasshook__`, explicit registering will be redundant.

File: Objects/dictobject.c
Comment (methane) [inline Objects/dictobject.c]: Three `dictreviter_iternextXXX` have most same code.
Since reviter is very rarely used compared to normal iter, please optimize for maintenance cost than execution speed.

Comment (remilapeyre) [reply Objects/dictobject.c]: Indeed thanks, I removed the useless functions.

File: Objects/dictobject.c
Comment (methane) [inline Objects/dictobject.c]: Can we use switch-case here?

File: Objects/dictobject.c
Comment (methane) [inline Objects/dictobject.c:3353]: This if statements looks ugly.

```
di->di_pos = 0;

if ((itertype == &PyDictRevIterKey_Type ||
     itertype == &PyDictRevIterItem_Type ||
     itertype == & PyDictRevIterValue_Type) && di->ma_used > 0)
{
    di->di_pos = dict->ma_keys->dk_nentries - 1;
}
```

This looks better, IMO.

This is matter of taste.  If you prefer current code, no need to rewrite.

**Review by methane — COMMENTED** [review]
You need to update document and what's new too.

File: Objects/dictobject.c
Comment (remilapeyre) [reply Objects/dictobject.c:3353]: This is indeed far better, thanks for noticing that

**Comment by remilapeyre** [comment]
> You need to update document and what's new too.

For What's New, is https://github.com/python/cpython/blob/71c552a38d5d0212cf5217443ffda00cfc48c3a7/Misc/NEWS.d/next/Core%20and%20Builtins/2018-05-23-17-18-02.bpo-33462.gurbpbrhe.rst not enough?

File: Objects/dictobject.c
Comment (remilapeyre) [reply Objects/dictobject.c]: I'm not sure it's possible don't `case`s needs to be int consts?

I tried and got this output from the compiler: `error: expression is not an integer constant expression`.

Comment (methane) [reply Objects/dictobject.c]: You're right.

**Comment by methane** [comment]
NEWS and What's New is different.
Compare https://docs.python.org/3/whatsnew/3.7.html and https://docs.python.org/3/whatsnew/changelog.html#python-3-7-0-final

NEWS contains most of all changes including small bugfixes.
"What's New" contains only important changes, like changing builtin APIs.

**Comment by remilapeyre** [comment]
I think the documentation could build on 639b43d, should I cherry-pick it, rebase the branch on master or something else?

**Comment by merwok** [comment]
Merging upstream master is the easiest option here (and better than rebase to avoid PR comments to lose their context).

File: Doc/library/stdtypes.rst
Comment (merwok) [inline Doc/library/stdtypes.rst]: A shortcut for ``reversed(d.keys())`` you mean?

Comment (remilapeyre) [reply Doc/library/stdtypes.rst]: Indeed, I will fix this.

I saw that many tests are not passing anymore, I must have made something wrong during the merge. I will add a patch.

Comment (remilapeyre) [reply Doc/library/stdtypes.rst]: @merwok Is the documentation good enough once this mistake will be fixed?

File: Doc/library/stdtypes.rst
Comment (merwok) [inline Doc/library/stdtypes.rst]: I think I would keep this text as it was before, and add a new short block to say “Dictionaries are reversible” and show the code examples.

**Comment by remilapeyre** [comment]
@methane it seems that 71c552 introduced some very strange behavior that I don't fully understand: when calling `__reduce__` on a reverse iterator, it raises `AttributeError` unless I called `help` on the object before. This broke pickling test, here is an excerpt:

    ➜  cpython git:(master) ./python.exe
    Python 3.8.0a0 (heads/master:d1d0e4d724, Sep 10 2018, 01:24:22) 
    [Clang 9.0.0 (clang-900.0.39.2)] on darwin
    Type "help", "copyright", "credits" or "license" for more information.
    >>> d = dict(a=1, b=2)
    >>> reversed(d.keys()).__reduce__
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    AttributeError: 'dict_reversekeyiterator' object has no attribute '__reduce__'
    >>> help(reversed(d.keys()))

    >>> reversed(d.keys()).__reduce__
    <built-in method __reduce__ of dict_reversekeyiterator object at 0x10c41f988>

Is there a reason for this behavior? I don't know Pickle well but this is very strange.

**Comment by methane** [comment]
@remilapeyre  You need to call PyType_READY() for new types.  See here:
https://github.com/python/cpython/blob/1f36bf6077d93cb43fd84bea4a8a625fa772d1fa/Objects/object.c#L1744-L1754

**Comment by remilapeyre** [comment]
Reading the documentation, I mistakingly understood that `PyType_Ready` would be called automatically, this is far more logical.

File: Doc/library/stdtypes.rst
Comment (methane) [inline Doc/library/stdtypes.rst:4292]: This section should be right after `describe:: iter(d)`.
``versionadded::`` is also helpful.

File: Lib/test/test_dict.py
Comment (methane) [inline Lib/test/test_dict.py]: Why `sorted`?  Can't we use `list`?

**Comment by methane** [comment]
Excuse me, I don't have time to final review before merge in this month.

File: Doc/library/stdtypes.rst
Comment (remilapeyre) [reply Doc/library/stdtypes.rst:4292]: Looking at other sections of the documentation, I thought it was supposed to be ordered alphabetically, isn't that so?

File: Lib/test/test_dict.py
Comment (remilapeyre) [reply Lib/test/test_dict.py]: I look at the other tests like `test_valuesiterator_pickling` to write my own, they were certainly written before dict were ordered.

It would indeed be probably better to test with `list`, after all, we do respect order now.
I will change the tests I've written. Do you want me to change the other tests as well or prefer not to?

**Review by remilapeyre — COMMENTED** [review]
That's not a problem, after all I can live with `reversed(iter(d.values()))`. Thanks for your help so far.

**Comment by remilapeyre** [comment]
@methane Would the PR be acceptable with the two changes you asked for?

File: Doc/library/stdtypes.rst
Comment (methane) [inline Doc/library/stdtypes.rst:4291]: `an`? `dictionnary`?

File: Doc/library/stdtypes.rst
Comment (methane) [inline Doc/library/stdtypes.rst]: No need for adding examples here.

File: Lib/test/test_dict.py
Comment (methane) [reply Lib/test/test_dict.py]: If you can.  Not mandatory.

File: Doc/library/stdtypes.rst
Comment (methane) [inline Doc/library/stdtypes.rst]: This versionchanged tag seems duplicated.  Remove `and dictionary views`.

File: Doc/library/stdtypes.rst
Comment (methane) [inline Doc/library/stdtypes.rst]: `d["one"] = 42` is example for "update value for a key doesn't change insertion order of a key".
It is not related reversed.  Remove these line and keep example minimal.
People don't read long example.

File: Objects/dictobject.c
Comment (methane) [inline Objects/dictobject.c]: Remove this empty line

File: Objects/dictobject.c
Comment (methane) [inline Objects/dictobject.c]:
```
}
else {
```

**Review by methane — CHANGES_REQUESTED** [review]

**Comment by remilapeyre** [comment]
I have made the requested changes; please review again

**Review by methane — APPROVED** [review]

**Comment by methane** [comment]
Please resolve conflict.

**Comment by remilapeyre** [comment]
@methane the conflicts should now be resolved

**Pull request merged**
